### PR TITLE
Daft Native Reader for Parquet Content Types 

### PR DESCRIPTION
**Pull request by samster25**
* Read Parquet using using daft's native reader for s3 files

### CODE REVIEW

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:277]:
```suggestion
        if "schema" in kwargs:
            daft_schema = pyarrow_to_daft_schema(kwargs["schema"])
            return  table.cast_to_schema(daft_schema).to_arrow()
        else:
            return table
```

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:265]:
```suggestion
        kwargs = {}
        if pa_read_func_kwargs_provider is not None:
            kwargs = pa_read_func_kwargs_provider(content_type, kwargs)
            
        table = Table.read_parquet(
            path=s3_url, 
            columns=column_names, 
            coerce_int96_timestamp_unit=TimeUnit.from_str(kwargs.get("coerce_int96_timestamp_unit", "ms")), 
```

**Review by raghumdani — COMMENTED** [review]
Thanks for the PR. I have few suggestions.

**Comment by raghumdani** [comment]
Please bump up the deltacat version so that we can publish it to PyPi as it is required for deployment.

File: deltacat/utils/pyarrow.py
Comment (samster25) [reply deltacat/utils/pyarrow.py:277]: Good call, done

File: deltacat/utils/pyarrow.py
Comment (samster25) [reply deltacat/utils/pyarrow.py]: done!

**Comment by samster25** [comment]
> Please bump up the deltacat version so that we can publish it to PyPi as it is required for deployment.

done!

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:279]: The key name is `session_token`.

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:273]: you may want to use `include_columns` here instead of columns as columns will have all columns while `include_columns` will only have columns to include.

Comment (samster25) [reply deltacat/utils/pyarrow.py]: done!

File: deltacat/utils/daft.py
Comment (raghumdani) [inline deltacat/utils/daft.py:9]: Can we address this TODO before merging into DeltaCAT?

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:259]: This can be abstracted to a separate method.

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:282]: Can we allow passing rest of the kwargs to `read_parquet` directly, so that we can modify io_config etc. without requiring a DeltaCAT change?

**Review by raghumdani — COMMENTED** [review]
Looks good overall. Please take a look at final set of suggestions.

**Comment by raghumdani** [comment]
Reduce the logging for S3 IO parquet reads.

File: requirements.txt
Comment (pdames) [inline requirements.txt:4]: Should this now be `getdaft==0.1.12`? Also, I think we can remove the corresponding line to install daft from https://github.com/ray-project/deltacat/blob/main/benchmark-requirements.txt#L4C11-L4C11 since it should be transitively included once this change is merged.

File: deltacat/utils/daft.py
Comment (jaychia) [reply deltacat/utils/daft.py]: Addressed!

File: requirements.txt
Comment (jaychia) [reply requirements.txt]: Done, and done!

File: deltacat/utils/pyarrow.py
Comment (samster25) [reply deltacat/utils/pyarrow.py]: Factored out this code into a separate function in `util.daft`

File: deltacat/utils/daft.py
Comment (raghumdani) [inline deltacat/utils/daft.py:52]:
```suggestion
    if kwargs.get("schema") is not None:
```

Just to be safe, when clients pass None value.

**Review by raghumdani — COMMENTED** [review]
Looks good overall. Can you also add unit tests (testing schema override as well) and remove the [Do not merge] from PR description?

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py:39]: The default behavior for daft is that it will read all of the columns from the parquet file. It looks like the current behavior for pyarrow also just uses include columns: https://github.com/ray-project/deltacat/blob/main/deltacat/utils/pyarrow.py#L231

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py:69]: it looks like it does both actually
image

Comment (samster25) [reply deltacat/utils/daft.py:69]: added a unit test anyway though

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py]: done

File: deltacat/utils/pyarrow.py
Comment (samster25) [reply deltacat/utils/pyarrow.py]: done!

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py]: done!

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py]: done

File: deltacat/tests/utils/test_daft.py
Comment (raghumdani) [inline deltacat/tests/utils/test_daft.py:7]: This test will start failing when you remove this S3 file from that location. Can you use a local file instead?

File: deltacat/tests/utils/test_daft.py
Comment (raghumdani) [inline deltacat/tests/utils/test_daft.py:18]: We may need another test with schema kwarg to make sure it's not adding null columns.

File: deltacat/utils/daft.py
Comment (raghumdani) [reply deltacat/utils/daft.py:59]: We recently changed the behavior, if column_names is also None, we can read all columns.

**Review by raghumdani — COMMENTED** [review]
LGTM!

File: deltacat/utils/daft.py
Comment (samster25) [reply deltacat/utils/daft.py]: fixed!

File: deltacat/tests/utils/test_daft.py
Comment (samster25) [reply deltacat/tests/utils/test_daft.py]: fixed!

File: deltacat/tests/utils/test_daft.py
Comment (samster25) [reply deltacat/tests/utils/test_daft.py:22]: added

File: deltacat/utils/pyarrow.py
Comment (samster25) [reply deltacat/utils/pyarrow.py:267]: done

**Review by raghumdani — APPROVED** [review]
LGTM! Thanks for this contribution.

**Comment by raghumdani** [comment]
Based on the latest job run, we are seeing lot of retries happening during materialize step of compaction. The diff is basically, 0.1.13 and the new commits in this PR. We can keep this PR open until this issue is resolved.

File: deltacat/utils/pyarrow.py
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:390]: Can we check "reader_type: ["daft", "pyarrow"]" in the kwargs returned by  `pa_read_func_kwargs_provider(content_type, {})`?

Comment (samster25) [reply deltacat/utils/pyarrow.py]: Went ahead and updated it. let me know if it's what you had in mind?

**Review by raghumdani — APPROVED** [review]
Thanks for this contribution.

**Comment by samster25** [comment]
> Thanks for this contribution.

No Problem! It was a blast working on it. btw, It doesn't look like I have permission to merge it, so you have to do it.

**Pull request merged**
